### PR TITLE
[core] Convert memory monitor cgroup values from ray check to log

### LINKED ISSUE #28579: [Core] memory_monitor crash
### What happened + What you expected to happen

https://console.anyscale-staging.com/o/anyscale-internal/projects/prj_qC3ZfndQWYYjx2cz8KWGNUL4/clusters/ses_GASWyegwPsGW8V2H2iuFvt88?command-history-section=command_history

(raylet) [2022-09-16 09:27:25,719 C 236 236] (raylet) memory_monitor.cc:121:  Check failed: (_left_ > _right_)  117896851456 vs 117896851456
(raylet) *** StackTrace Information ***
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x485f6a) [0x55b0b8ce2f6a] ray::operator<<()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x487a42) [0x55b0b8ce4a42] ray::SpdLogMessage::Flush()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x487d57) [0x55b0b8ce4d57] ray::RayLog::~RayLog()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x45cfc3) [0x55b0b8cb9fc3] ray::MemoryMonitor::GetCGroupMemoryBytes()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x45ed9b) [0x55b0b8cbbd9b] ray::MemoryMonitor::GetMemoryBytes()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x45edfc) [0x55b0b8cbbdfc] std::_Function_handler<>::_M_invoke()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x4126bb) [0x55b0b8c6f6bb] ray::PeriodicalRunner::DoRunFnPeriodicallyInstrumented()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x413e10) [0x55b0b8c70e10] std::_Function_handler<>::_M_invoke()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x469da6) [0x55b0b8cc6da6] EventTracker::RecordExecution()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x41069e) [0x55b0b8c6d69e] ray::PeriodicalRunner::DoRunFnPeriodicallyInstrumented()::{lambda()#1}::operator()()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x41157f) [0x55b0b8c6e57f] boost::asio::detail::wait_handler<>::do_complete()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x994b6b) [0x55b0b91f1b6b] boost::asio::detail::scheduler::do_run_one()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x996331) [0x55b0b91f3331] boost::asio::detail::scheduler::run()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x996560) [0x55b0b91f3560] boost::asio::io_context::run()
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x140ae4) [0x55b0b899dae4] main
(raylet) /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f61a991e083] __libc_start_main
(raylet) /home/ray/anaconda3/lib/python3.7/site-packages/ray/core/src/ray/raylet/raylet(+0x17d777) [0x55b0b89da777]
(raylet) 

### Versions / Dependencies

latest

### Reproduction script

n/a

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by clarng**
Signed-off-by: Clarence Ng <clarence.wyng@gmail.com>

## Why are these changes needed?

We see ray checks triggered in Anyscale and crash raylet. Why this is happening:
- Anyscale started setting memory limit in its container
- Its observed that the container memory usage in cgroup v2 can be equal or greater than limit, which indicates some level of short-term bursting that is allowed by the cgroup, i.e. the limit is not a hard cap. And in the reported case the usage is exactly equal to limit, but that is using cgroup v1

- it means neither the cgroup nor our memory monitor can guarantee the memory usage is below the OS limit

Therefore, it is safer to log instead of crash the raylet if the cgroup usage is equal to or above limit. 

Also convert usage > 0 check to always setting the usage to be positive and log a warning

## Related issue number

Closes https://github.com/ray-project/ray/issues/28579

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

117896851456

### CODE REVIEW

File: src/ray/common/memory_monitor.cc [hunk 1]
```diff
@@ -118,7 +118,17 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetCGroupMemoryBytes() {
             (total_bytes != kNull && used_bytes != kNull));
   if (total_bytes != kNull) {
     RAY_CHECK_GT(used_bytes, 0);
-    RAY_CHECK_GT(total_bytes, used_bytes);
+    if (used_bytes <= 0) {
+      RAY_LOG_EVERY_MS(WARNING, kLogIntervalMs)
```
Comment (scv119) [inline src/ray/common/memory_monitor.cc:122]: in what circumstance the used_bytes will be negative?

File: src/ray/common/memory_monitor.cc [hunk 2]
```diff
@@ -118,7 +118,17 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetCGroupMemoryBytes() {
...
+          << " Memory usage from cgroup is less than or equal to zero: " << used_bytes;
+    }
+    if (used_bytes >= total_bytes) {
+      RAY_LOG_EVERY_MS(WARNING, kLogIntervalMs)
+          << " Used memory is less than or equal to total memory used. This can "
```
Comment (scv119) [inline src/ray/common/memory_monitor.cc:127]: greater or equal to?

Comment (clarng) [reply src/ray/common/memory_monitor.cc]: Fixed, PTAL

File: src/ray/common/memory_monitor.cc (hunk 1 above)
Comment (clarng) [reply src/ray/common/memory_monitor.cc]: if we can't read / read something weird from the OS?
I changed so log and set used to 0 in these cases

**Review by scv119 — APPROVED** [review]

File: src/ray/common/memory_monitor.cc [hunk 3]
```diff
@@ -116,9 +116,19 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetCGroupMemoryBytes() {
...
             (total_bytes != kNull && used_bytes != kNull));
+  if (used_bytes < 0) {
+    RAY_LOG_EVERY_MS(WARNING, kLogIntervalMs)
+        << "Got negative used memory for cgroup " << used_bytes << ", setting it to zero";
+    used_bytes = 0;
+  }
   if (total_bytes != kNull) {
-    RAY_CHECK_GT(used_bytes, 0);
-    RAY_CHECK_GT(total_bytes, used_bytes);
+    if (used_bytes >= total_bytes) {
```
Comment (mwtian) [inline src/ray/common/memory_monitor.cc:125]: Should we set used_bytes to total_bytes in this case too?

**Review by mwtian — APPROVED** [review]

File: src/ray/common/memory_monitor.cc (hunk 3 above)
Comment (clarng) [reply src/ray/common/memory_monitor.cc:125]: Sure

**Pull request merged**